consensus/docs: draft Genesis design document for review (#3546)
input-output-hk/ouroboros-network: ouroboros-consensus/docs/GenesisDesignOverview.md

    from disk before realizing you don't need it.)

  * Mitigation: Immediately ignore blocks that are older than `k`. Checked via
    the ~1100 byte block header. Ouroboros papers' Common Prefix theorems
    ensure this is sound for nodes that have been "caught up" for long enough.

  * (TODO Confirm.) We disconnect from peer if they send a block that doesn't
    match the header they used to convince us to download the block.

2) "long range" attacks (Genesis Paper)

  * With just Praos, the adversary can isolate any node that hasn't been
    "caught up" to `mainnet` _recently_ enough, such as an entirely fresh node,
    a node that's been offline a month, etc.

  * Current code hasn't worried about this because our deployment instructs
    people to sync from one of the IOG relays, which have forever been "caught
    up".

  * We want to shutdown the relays, which will arm the long range attack. So we
    need to mitigate directly.

  * See below for more details dynamics and the mitigation scheme.

3) "eclipse" attack (no IOG paper for this)

  * Even with Genesis, if all our peers are adversarial, we're doomed. The

>     Genesis paper can kind of get away with it, because it rollsback by more

Reviewer: ```suggestion
    Genesis paper can kind of get away with it because it rolls back by more
```
also *it*?

Author: Done, thanks.

honest chain. On the other hand, the longest prefix a node has received at any
given time might just be the one that node happened to receive first. Thus the
Genesis paper (via Edsko's interpretation) instead defines _prefix selection_,
which is to prefer the prefix that is denser in fixed-size window of slots that
starts just after the intersection with the competing candidate prefix --
called _the (comparison's) Genesis window_.

> Given a large enough window starting at two chain's intersection, the honest
chain should always be densest, for effectively the same reasons that the
honest chain should always be longest in the system.

Prefix selection alone doesn't prevent long range attacks, though, since
there's no guarantee that a node has even been sent an honest prefix! For that
concern, we've decided to play a numbers game: trust in the P2P peer selection
algorithm insofar as to assume that if we have ~20 peers, then at least one of
them is feeding us (prefixes of) the honest chain. If that's not true, then
we've been eclipsed (likely a botnet attack).

## Another reason the intersection matters

The above prefix selection perspective on the Genesis rule is particularly

> pertinent for our code because our actual node, in contrast to the the model in

Reviewer: ```suggestion
pertinent for our code because our actual node, in contrast to the model in
```

Author: Done, thanks.

the papers, exchanges blocks at a time instead of whole chains, and hence, when
(re)joining, is comparing prefixes until catches back up to `mainnet`. However,
even if we did (somehow!) exchange entire chains, the Genesis rule is still
necessary.

Given enough wallclock time and enough legitimate leadership opportunities, an
adversary could craft a chain (TODO eg via brute-force search/grinding?) that
allots them a larger and larger share of the active slots. In various ways,
such a chain could actually become longer than the honest chain. That would not
fool the Genesis rule, though, since such a chain would be comparatively sparse
immediately after the intersection, because the Genesis window is so short that
the adversary would not have enough elections to manipulate the leader schedule
within it.

In summary, without the Genesis rule, total chain length might be an
apples-to-oranges, since, for historical chains the adversary has had

> sufficient wallclock time and leadership opporutnities to manipulate the chain

Reviewer: ```suggestion
sufficient wallclock time and leadership opportunities to manipulate the chain
```

Author: Done, thanks.

     Bests_i = the (necessarily non-empty) subset of Candidates_i
               containing each prefix for which
               the count of headers in the Genesis window starting at Anchor_i
               plus the count of slots that are
                   in the Genesis window starting at Anchor_i
                   > the slot of the prefix's tip
                   and <= the slot of the peer's claimed tip
               is >= Cutoff_i

TODO would it be clearer to map each prefix-and-claimed-tip pair to an interval
[#headers in window, min(anchor+s, claimed tip) - #empty slots in window older
than a header] and keep only those prefixes whose upper bound is >= the overall
minimum lower bound?

```

Intuitively, an inferior candidate prefix is eliminated from consideration (and
thereby stops holding back `Anchor_j`) in one of two ways. Either that chain
ends in the window (while the better chains don't) or else the node eventually
receives enough of each prefix in the window to see that the number of
remaining slots in the window for this sparse prefix is too small for the
prefix to ever be able to catch up to the greatest number of received headers

> in the window on whichever prefix has the most.

Reviewer 2: is this how we ensure we've received "enough" headers?

I'm not clear on what the threshold is for prefix selection that makes it confident it's got enough headers to make a decision...

vastly increases the chance that their chain remains in the `Bests` set, which
in turn causes `Anchor_j` to be shorter than it could have been. They would
thus prevent a syncing node from selecting any more of the honest chain, since
prefix selection wouldn't be able to choose between their offering and the
honest chain.

* Their selection has a low block count, but a high number of remaining slots.

* The honest chain has a much higher block count, but a low number of remaining
  slots.

* Since both are about the lowest options block count, both will remain in
  `Bests`.

For this reason, while in Syncing, a node must require that the headers sent by
each peer improve over time. EG on average (ie leaky token bucket), every 10
seconds or so they have sent a header with a higher block number than any they
had sent before. If a peer can't even maintain that average for ~1100 byte
messages, then they're probably not actually sending me a real historical chain
(ie one that does not require any ChainSync `MsgRollback`s). Or their network
connection is so poor that I should replace them with a new peer anyway.

# Eclipse detection

> That description of constitutes the expected path through the Genesis state

Reviewer: description of?

Author: Done, thanks.

properly configured nodes, that remain in CaughtUp enough of the time, and so
on). If that's always true, then (prefixes of) the honest chain will win every
density comparison forever.

## Detection scenarios

If I'm eclipsed, then I won't see (prefixes of) the honest chain. How can I
notice? We consider two cases.

* For a fresh node that's joining the network for the first time -- one that
  was never before CaughtUp -- the only hope for detecting eclipse is for the
  chain you're fed to be objectively too sparse. EG a chain that has blocks in
  less than 40% of the active slots that that chain's ledger states' leadership
  schedules identify as active. But this rule would backfire catastrophically
  if Cardano mainnet ever reached (near) that threshold, thus this rule would
  be introducing a new failure mode if enough honest stakeholders gave up
  Cardano (without unregistering their stake) even in the absence of an
  adversary.

* For a node that's rejoining the network, there's another option, since it
  already has selected a prefix of the (presumably) honest chain. If we were to
  become eclipsed by an adversary, which can never have more than half of the

>   active stake, then -- given a "long enough" slot interval on either side --

Reviewer: Either side of?

Author: Done, thanks.

  we should see a portion of our recent chain when the density plummeted.

We could instead collapse those two by defining the density up to the genesis
block to be two times whatever the absolute cutoff is. So I'll only discuss the
rejoining case from now on.

> ## Balancing false positives and missed alarms

Reviewer 2: Good document overall. Reminder to include all outputs from discussion with Alexander Russell in this section...

The key challenge is false positives -- a node wrongly concluding it's
eclipsed. This possibility is unavoidable, since we're working with partial
information, much of which is random. As a drastic example: there is a
probability that some epoch has no active slots, an empty leader schedule! That
probability is astronomically small, but not zero. So we have to balance the
sensitivity of our detection against its robustness, false alarms versus missed
alarms. TODO also its promptness: we'd ideally detect an eclipse sooner rather
than later... is a slow alarm just a partially/temporarily missed alarm?

We can't judge that balance well without understanding the costs of each.

* A missed alarm means some nodes were eclipsed and did not notice. In this
  case the eclipsed nodes suffer the direct consequence, such as selling a
  yacht without receiving payment on the honest chain.

  The network as a whole wouldn't really notice the absence of a few nodes'
  stake, which is inactive while they're eclipsed. On the other hand, it could

>   be distastrous if a large portion of the network were eclipsed simultaneously

Reviewer: ```suggestion
  be disastrous if a large portion of the network were eclipsed simultaneously
```

Author: Done, thanks.

  wasn't. The direct consequence is that the node leaves CaughtUp and goes back
  to Connecting. If we put enough defensive mechanisms in place in Connecting
  and Syncing (mostly timeouts), then such a (non-eclipsed!) node will take at
  most a few minutes to return to CaughtUp (each false alarm).

  A node in Connecting or Syncing is more likely to waste its leadership
  opportunities by minting blocks that do not extend the best chain in the
  network. This is because no growth (in Connecting) and Genesis prefix
  selection (in Syncing) is slower to select that chain than is Praos chain
  selection (in CaughtUp).

  Again, the network as a whole wouldn't notice the absence of a few nodes'
  stake, but it would be disastrous if the adversary could induce false alarms
  in significant portions of the honest stake with appreciable
  duration/frequency/duty cycle.

## Why Genesis prefix selection can be slower than Praos chain selection

It was decided the node should still mint blocks when in Connecting and
Syncing. Any honest nodes minting stale blocks (creating several short forks)
could slightly help the adversary create alternative chains, but the risk of

> many honest nodes skipping many slots they could have created a competitive is

Reviewer: It seems there's something missing in this sentence :thinking:

Reviewer: Chain?

Author: Done, thanks.

TODO if "recent" is too long of a duration, then we won't detect eclipses very
quickly

TODO False alarms may harm the eclipsed node because it might mint a
non-competitive block. If that happens to too many nodes/too often, then it
lowers the honest chain growth.

TODO Missed alarms harm the eclipsed node, because 1) it mints non-competitve
blocks while eclipsed and 2) the table of settlement times are misleading during
an eclipse (eg a yacht merchant could get bamboozled). If an eclipse lasts long
enough, then the node can't rejoin the honest network without intervention. If
too many nodes are eclipsed at once (or for so long they can't recover). If that
happens to too many nodes/too often, then it lowers the honest chain growth.

TODO False alarms inherently cause harm in the absence of attacks, but might
potentially be easy enough for an adversary to trigger that they become an
attack vector of their own!

# BlockFetch in CaughtUp

TODO download capacity could be a limiting factor: so which blocks to download
when faced with competing chains? The Tse paper
https://eprint.iacr.org/2021/1545 says to download blocks along the chain with
the freshest block.

> TODO The current code does a hybrid of prefering the longest blocks offered by

Reviewer: ```suggestion
TODO The current code does a hybrid of preferring the longest blocks offered by
```

Author: Done, thanks.
